modules/platforms/cpp/common/src/common/thread_pool.cpp

```diff
@@ -0,0 +1,156 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <ignite/common/thread_pool.h>
+
+namespace ignite
+{
+    namespace common
+    {
+        ThreadPool::ThreadPool(uint32_t threadsNum):
+            started(false),
+            stopped(false),
+            threads()
+        {
+            uint32_t threadToStart = threadsNum != 0 ? threadsNum : concurrent::GetNumberOfProcessors();
+            if (!threadToStart)
+                threadToStart = 2;
+
+            threads.reserve(threadToStart);
+            for (uint32_t i = 0; i < threadToStart; ++i)
+            {
+                SP_WorkerThread thread(new WorkerThread(queue));
+                threads.push_back(thread);
+            }
+        }
+
+        ThreadPool::~ThreadPool()
+        {
+            Stop();
+        }
+
+        void ThreadPool::Start()
+        {
+            if (started)
+                return;
+
+            started = true;
+
+            for (std::vector<SP_WorkerThread>::iterator it = threads.begin(); it != threads.end(); ++it)
+                it->Get()->Start();
+        }
+
+        void ThreadPool::Stop()
+        {
+            if (stopped)
+                return;
+
+            stopped = true;
+
+            queue.Unblock();
+
+            for (std::vector<SP_WorkerThread>::iterator it = threads.begin(); it != threads.end(); ++it)
+                it->Get()->Join();
+        }
+
+        void ThreadPool::Dispatch(const SP_ThreadPoolTask& task)
+        {
+            queue.Push(task);
+        }
+
+        ThreadPool::TaskQueue::TaskQueue() :
+            unblocked(false)
+        {
+            // No-op.
+        }
+
+        ThreadPool::TaskQueue::~TaskQueue()
+        {
+            // No-op.
+        }
+
+        void ThreadPool::TaskQueue::Push(const SP_ThreadPoolTask &task)
+        {
+            if (!task.IsValid())
+                return;
+
+            concurrent::CsLockGuard guard(mutex);
+
+            tasks.push_back(task);
+
+            waitPoint.NotifyOne();
+        }
+
+        SP_ThreadPoolTask ThreadPool::TaskQueue::Pull()
+        {
+            concurrent::CsLockGuard guard(mutex);
+            if (unblocked)
+                return SP_ThreadPoolTask();
+
+            while (tasks.empty())
+            {
+                waitPoint.Wait(mutex);
+
+                if (unblocked)
+                    return SP_ThreadPoolTask();
+            }
+
+            SP_ThreadPoolTask res = tasks.front();
+            tasks.pop_front();
+            return res;
+        }
+
+        void ThreadPool::TaskQueue::Unblock()
+        {
+            concurrent::CsLockGuard guard(mutex);
```

Comment:
Igor, is it ok that tasks when thread-pool is stopped are abandoned?
May be it is good idea to complete task exceptionally?

Reply:
Will handle when will fix issue that @ptupitsyn suggested.